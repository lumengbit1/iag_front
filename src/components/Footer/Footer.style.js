import styled from 'styled-components';
import NumberFormat from 'react-number-format';

export const Root = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  width: 100%;
  font-size: 28px;
  font-weight: bold;
  position: fixed;
  bottom: 0;
  left: 0;
  background-color: white;
  border-top: solid 1px black;
  padding: 20px 0;
`;

export const Input = styled(NumberFormat)`
  text-align: center;
  width: 200px;
  height: 40px;
  border: solid 1px black;
  box-sizing: border-box;
  font-size: 24px;
  margin-bottom: 10px;
`;

export const Button = styled.button`
  width: 150px;
  height: 40px;

  font-size: 24px;

  &:hover {
    cursor: pointer;
  }

  &:first-child {
    margin-right: 20px;
  }
`;

export const ButtonContainer = styled.div`
  display: flex;
`;

export const Text = styled.div`
  color: red;
  text-transform: uppercase;
`;
