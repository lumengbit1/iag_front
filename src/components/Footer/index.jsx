import React from 'react';
import ImmutablePropTypes from 'react-immutable-proptypes';
// import { useImmer } from 'use-immer';
import { useDispatch } from 'react-redux';
import _ from 'lodash';
import { postGuessAction, resetAction } from '../../reducers/actions';
import {
  Root,
  Input,
  Button,
  ButtonContainer,
  Text,
} from './Footer.style';

const Footer = (props) => {
  const dispatch = useDispatch();
  const [inputValue, setInputValue] = React.useState('');

  const isCorrect = !props.reponseValue.isEmpty() && props.reponseValue.last().get('correct');

  const reset = () => {
    dispatch(resetAction());
    setInputValue('');
  };

  const submit = () => {
    dispatch(postGuessAction(_.join(props.hintValue.toJS(), ''), inputValue));
    setInputValue('');
  };

  return (
    <Root>
      <Input
        data-testid="input"
        onChange={(event) => setInputValue(event.target.value)}
        value={inputValue}
        maxLength="8"
        placeholder="Type Here"
      />
      {isCorrect && (
        <Text>
          Congratulations!
        </Text>
      )}
      <ButtonContainer>
        <Button
          data-testid="submit"
          onClick={submit}
          disabled={_.size(inputValue) < 8}
        >
          Submit
        </Button>

        <Button
          data-testid="reset"
          onClick={reset}
        >
          Reset
        </Button>
      </ButtonContainer>
    </Root>
  );
};

Footer.propTypes = {
  hintValue: ImmutablePropTypes.list.isRequired,
  reponseValue: ImmutablePropTypes.list.isRequired,
};

export default Footer;
