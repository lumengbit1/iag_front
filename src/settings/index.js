export default {
  RESULTS_BASE_API_DOMAIN: 'http://localhost:3000/results',
  SAVED_BASE_API_DOMAIN: 'http://localhost:3000/saved',
};
