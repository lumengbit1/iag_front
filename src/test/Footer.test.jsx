import React, { useState as useStateMock } from 'react';
import 'jest-styled-components';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';
import { combineReducers } from 'redux-immutable';
import { fromJS } from 'immutable';
import * as redux from 'react-redux';
import { createStore, applyMiddleware } from 'redux';
import axios from 'axios';
import MockAdapter from 'axios-mock-adapter';
import { render, cleanup, waitFor, fireEvent } from '@testing-library/react';
import Footer from '../components/Footer';
import getReducer from '../reducers/get';
import postReducer from '../reducers/post';
import settings from '../settings';

const store = createStore(combineReducers({ getReducer, postReducer }), applyMiddleware(thunk));

jest.mock('react', () => ({
  ...jest.requireActual('react'),
  useState: jest.fn(),
}));

const mockData = fromJS([
  {
    correct: false,
    highlight: [],
    hint: 'test',
    answer: '1111',
  },
  {
    correct: false,
    highlight: [1],
    hint: 'test',
    answer: '2222',
  },
]);

const mockhint = fromJS([
  {
    correct: false,
    highlight: [],
    hint: 'test',
    answer: '1111',
  },
]);

afterEach(cleanup);

describe('Render Test', () => {
  const setInputValue = jest.fn();

  beforeEach(() => {
    useStateMock.mockImplementation((inputValue) => [inputValue, setInputValue]);
  });

  it('1: expect rendering correct', () => {
    const { container } = render(
      <Provider store={store}>
        <Footer reponseValue={mockData} hintValue={mockhint} />
      </Provider>,
    );

    expect(container).toMatchSnapshot();
  });

  it('2: expect reset button onclick event', async () => {
    const mock = new MockAdapter(axios);
    mock.onGet(`${settings.BASE_API_DOMAIN}/reset`).reply(200, mockData);
    const useDispatchSpy = jest.spyOn(redux, 'useDispatch');
    const mockDispatchFn = jest.fn();
    useDispatchSpy.mockReturnValue(mockDispatchFn);
    const { getByTestId } = render(
      <redux.Provider store={store}>
        <Footer reponseValue={mockData} hintValue={mockhint} />
      </redux.Provider>,
    );
    await waitFor(() => getByTestId('reset'));
    fireEvent.click(getByTestId('reset'));
    expect(mockDispatchFn).toBeCalled();
  });

  it('3: expect input onChange event', async () => {
    const onChangeMock = jest.fn();
    const event = {
      target: { value: '11111111' },
    };
    const { getByTestId } = render(
      <redux.Provider store={store}>
        <Footer reponseValue={mockData} onChange={onChangeMock} hintValue={mockhint} />
      </redux.Provider>,
    );
    await waitFor(() => getByTestId('input'));
    fireEvent.change(getByTestId('input'), event);
    expect(setInputValue).toBeCalledWith('11111111');
  });
});
