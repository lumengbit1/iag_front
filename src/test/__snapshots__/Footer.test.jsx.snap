// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render Test 1: expect rendering correct 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
  font-size: 28px;
  font-weight: bold;
  position: fixed;
  bottom: 0;
  left: 0;
  background-color: white;
  border-top: solid 1px black;
  padding: 20px 0;
}

.c1 {
  text-align: center;
  width: 200px;
  height: 40px;
  border: solid 1px black;
  box-sizing: border-box;
  font-size: 24px;
  margin-bottom: 10px;
}

.c3 {
  width: 150px;
  height: 40px;
  font-size: 24px;
}

.c3:hover {
  cursor: pointer;
}

.c3:first-child {
  margin-right: 20px;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

<div>
  <div
    class="c0"
  >
    <input
      class="c1"
      data-testid="input"
      inputmode="numeric"
      maxlength="8"
      placeholder="Type Here"
      type="text"
      value=""
    />
    <div
      class="c2"
    >
      <button
        class="c3"
        data-testid="submit"
        disabled=""
      >
        Submit
      </button>
      <button
        class="c3"
        data-testid="reset"
      >
        Reset
      </button>
    </div>
  </div>
</div>
`;
